errbot/flow.py

```diff
@@ -0,0 +1,335 @@
+                           automatically this flow with no user intervention.
+        :return: the newly created node if you passed a command or the node you gave it to be easily chainable.
+        """
+        node_to_connect_to = node_or_command if isinstance(node_or_command, FlowNode) else FlowNode(node_or_command)
+        self.children.append((predicate, node_to_connect_to))
+        return node_to_connect_to
+
+    def predicate_for_node(self, node: 'FlowNode'):
+        """
+        gets the predicate function for the specified child node.
+        :param node: the child node
+        :return: the predicate that allows the automatic execution of that node.
+        """
+        for predicate, possible_node in self.children:
+            if node == possible_node:
+                return predicate
+        return None
+
+    def __str__(self):
+        return self.command
+
+
+class FlowRoot(FlowNode):
+    """
+    This represent the entry point of a flow description.
+    """
+    def __init__(self, name: str, description: str):
+        """
+
+        :param name: The name of the conversation/flow.
+        :param description:  A human description of what this flow does.
+        """
+        super().__init__()
+        self.name = name
+        self.description = description
+        self.auto_triggers = []
+
+    def connect(self,
+                node_or_command: Union['FlowNode', str],
+                predicate: Predicate=lambda _: False,
+                auto_trigger: bool=False):
+        """
+        :see: FlowNode except fot auto_trigger
+        :param predicate: :see: FlowNode
+        :param node_or_command: :see: FlowNode
+        :param auto_trigger: Flag this root as autotriggering: it will start a flow if this command is executed
+                              in the chat.
+        """
+        resp = super().connect(node_or_command, predicate)
+        if auto_trigger:
+            self.auto_triggers.append(node_or_command)
+        return resp
+
+    def __str__(self):
+        return self.name
+
+
+class InvalidState(Exception):
+    """
+    Raised when the Flow Executor is asked to do something contrary to the contraints it has been given.
+    """
+    pass
+
+
+class Flow(object):
+    """
+    This is a live Flow. It keeps context of the conversation (requestor and context).
+    Context is just a python dictionary representing the state of the conversation.
+    """
+    def __init__(self, root: FlowRoot, requestor: Identifier, initial_context):
+        """
+
+        :param root: the root of this flow.
+        :param requestor: the user requesting this flow.
+        :param initial_context: any data we already have that could help executing this flow automatically.
+        """
+        self._root = root
+        self._current_step = self._root
+        self.ctx = dict(initial_context)
+        self.requestor = requestor
+
+    def next_autosteps(self) -> List[FlowNode]:
```

Comment:
as a client of the class you either want to know:

- what can I do next now ? (from executor)
- what is everything I can do  ? (from the UI to propose the user what he can do)